java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java

```diff
@@ -0,0 +1,279 @@
+    }
+  }
+
+  private static MethodInvocationMatcherCollection closeMethodInvocationMatcher() {
+    return MethodInvocationMatcherCollection.create(
+      MethodInvocationMatcher.create()
+        .typeDefinition(TypeCriteria.subtypeOf(JAVA_IO_CLOSEABLE))
+        .name(CLOSE_METHOD_NAME)
+        .withNoParameterConstraint(),
+      MethodInvocationMatcher.create()
+        .typeDefinition(TypeCriteria.subtypeOf(JAVA_LANG_AUTOCLOSEABLE))
+        .name(CLOSE_METHOD_NAME)
+        .withNoParameterConstraint());
+  }
+
+  private static boolean isCloseableOrAutoCloseableSubtype(Type type) {
+    return type.isSubtypeOf(JAVA_IO_CLOSEABLE) || type.isSubtypeOf(JAVA_LANG_AUTOCLOSEABLE);
+  }
+
+  private static boolean isIgnoredCloseableSubtype(Type type) {
+    for (String fullyQualifiedName : IGNORED_CLOSEABLE_SUBTYPES) {
+      if (type.isSubtypeOf(fullyQualifiedName)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private static boolean isSubclassOfInputStreamOrOutputStreamWithoutClose(Type type) {
+    Symbol.TypeSymbol typeSymbol = type.symbol();
+    Type superClass = typeSymbol.superClass();
+    if (superClass != null && (superClass.is("java.io.OutputStream") || superClass.is("java.io.InputStream"))) {
+      return typeSymbol.lookupSymbols(CLOSE_METHOD_NAME).isEmpty();
+    }
+    return false;
+  }
+
+  @Override
+  public void visitVariable(VariableTree tree) {
+    ExpressionTree initializer = tree.initializer();
+
+    // check first usage of closeables in order to manage use of same symbol
+    ignoreClosableSymbols(initializer);
+
+    Symbol symbol = tree.symbol();
+    if (isCloseableOrAutoCloseableSubtype(symbol.type())) {
+      executionState.newValueForSymbol(symbol, tree, getCloseableStateFromExpression(symbol, initializer));
+    }
+  }
+
+  private CloseableState getCloseableStateFromExpression(Symbol symbol, @Nullable ExpressionTree expression) {
+    if (shouldBeIgnored(symbol, expression)) {
+      return CloseableState.IGNORED;
+    } else if (isNull(expression)) {
+      return CloseableState.NULL;
+    } else if (expression.is(Tree.Kind.NEW_CLASS)) {
+      if (usesIgnoredCloseableAsArgument(((NewClassTree) expression).arguments())) {
+        return CloseableState.IGNORED;
+      }
+      return CloseableState.OPEN;
+    }
+    // TODO SONARJAVA-1029 : Engine currently ignore closeables which are retrieved from method calls. Handle them as OPEN.
+    return CloseableState.IGNORED;
+  }
+
+
+  private static boolean isNull(ExpressionTree expression) {
+    return expression == null || expression.is(Tree.Kind.NULL_LITERAL);
+  }
+
+  private  boolean shouldBeIgnored(Symbol symbol, @Nullable ExpressionTree expression) {
+    return shouldBeIgnored(symbol) || shouldBeIgnored(expression);
+  }
+
+  private boolean shouldBeIgnored(Symbol symbol) {
+    CloseableState state = (CloseableState) executionState.getStateOf(symbol);
+    return (state != null && state.isIgnored()) || symbol.isFinal()
```

Comment:
:new: ![MAJOR](https://raw.githubusercontent.com/henryju/image-hosting/master/major.png)Reduce the number of conditional operators (4) used in the expression (maximum allowed 3).[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1067)